hail/src/main/scala/is/hail/expr/ir/Parser.scala

```diff
@@ -0,0 +1,1009 @@
+        var r = in
+
+        val source = in.source
+        val offset = in.offset
+        val start = handleWhiteSpace(source, offset)
+        r = r.drop(start - offset)
+
+        if (r.atEnd || r.first != delim)
+          return Failure(s"consumed $what", r)
+        r = r.rest
+
+        val sb = new StringBuilder()
+
+        val escapeChars = "\\bfnrt'\"`".toSet
+        var continue = true
+        while (continue) {
+          if (r.atEnd)
+            return Failure(s"unterminated $what", r)
+          val c = r.first
+          r = r.rest
+          if (c == delim)
+            continue = false
+          else {
+            sb += c
+            if (c == '\\') {
+              if (r.atEnd)
+                return Failure(s"unterminated $what", r)
+              val d = r.first
+              if (!escapeChars.contains(d))
+                return Failure(s"invalid escape character in $what", r)
+              sb += d
+              r = r.rest
+            }
+          }
+        }
+        Success(unescapeString(sb.result()), r)
+      }
+    }
+
+  override def stringLiteral: Parser[String] =
+    quotedLiteral('"', "string literal") | quotedLiteral('\'', "string literal")
+
+  def backtickLiteral: Parser[String] = quotedLiteral('`', "backtick identifier")
+
+  def identifier = backtickLiteral | ident
+
+  def string_literal: Parser[String] = stringLiteral
+
+  def int64_literal: Parser[Long] = wholeNumber.map(_.toLong)
+
+  def float64_literal: Parser[Double] =
+      "-inf" ^^ { _ => Double.NegativeInfinity } |
+      """-?\d+(\.\d+)?[eE][+-]?\d+""".r ^^ { _.toDouble } |
+      """-?\d*\.\d+""".r ^^ { _.toDouble } // FIXME: Is this correct??
+
+  def parse(code: String): Array[Token] = {
+    parseAll(lexer, code) match {
+      case Success(result, _) => result
+      case NoSuccess(msg, next) => ParserUtils.error(next.pos, msg)
+    }
+  }
+}
+
+case class IRParserEnvironment(
+  refMap: Map[String, Type] = Map.empty,
+  irMap: Map[String, BaseIR] = Map.empty
+) {
+  def update(newRefMap: Map[String, Type] = Map.empty, newIRMap: Map[String, BaseIR] = Map.empty): IRParserEnvironment =
+    copy(refMap = refMap ++ newRefMap, irMap = irMap ++ newIRMap)
+
+  def withRefMap(newRefMap: Map[String, Type]): IRParserEnvironment = {
+    assert(refMap.isEmpty)
+    copy(refMap = newRefMap)
+  }
+
+  def +(t: (String, Type)): IRParserEnvironment = copy(refMap = refMap + t, irMap)
+}
+
+object IRParser {
+  def error(t: Token, msg: String): Nothing = ParserUtils.error(t.pos, msg)
+
+  def consumeToken(it: TokenIterator): Token = {
+    if (!it.hasNext)
+      fatal("No more tokens to consume.")
+    it.next()
+  }
+
+  def punctuation(it: TokenIterator, symbol: String): String = {
+    consumeToken(it) match {
+      case x: PunctuationToken if x.value == symbol => x.value
+      case x: Token => error(x, s"Expected punctuation with symbol '$symbol' but found ${ x.getClass } with value '${ x.value }'.")
```

Comment:
Just edit this a little.  Make it read "Expected punctuation '(' but found integer '57'".  Add a Token.getName function that returns the human readable name.